benches/engines/mod.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate criterion;
+extern crate kvproto;
+extern crate test_util;
+extern crate tikv;
+
+use criterion::{black_box, Bencher, Criterion};
+use kvproto::kvrpcpb::Context;
+use std::fmt;
+use test_util::generate_random_kvs;
+use tikv::storage::engine::{
+    BTreeEngine, Engine, Modify, RocksEngine, Snapshot, TestEngineBuilder,
+};
+use tikv::storage::{Key, CF_DEFAULT};
+
+const DEFAULT_KEY_LENGTH: usize = 64;
+const DEFAULT_ITERATIONS: usize = 1000;
+
+trait EngineFactory<E: Engine>: Clone + fmt::Debug + 'static {
+    fn build(&self) -> E;
+}
+
+#[derive(Clone)]
+struct BTreeEngineFactory {}
+
+impl EngineFactory<BTreeEngine> for BTreeEngineFactory {
+    fn build(&self) -> BTreeEngine {
+        BTreeEngine::default()
+    }
+}
+
+impl fmt::Debug for BTreeEngineFactory {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "BTreeEngine")
+    }
+}
+
+#[derive(Clone)]
+struct RocksEngineFactory {}
+
+impl EngineFactory<RocksEngine> for RocksEngineFactory {
+    fn build(&self) -> RocksEngine {
+        TestEngineBuilder::new().build().unwrap()
+    }
+}
+
+impl fmt::Debug for RocksEngineFactory {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "RocksEngine")
+    }
+}
+
+fn fill_engine_with<E: Engine>(engine: &E, expect_engine_keys_count: usize, value_length: usize) {
+    let mut modifies: Vec<Modify> = vec![];
+    if expect_engine_keys_count > 0 {
+        let kvs = generate_random_kvs(expect_engine_keys_count, DEFAULT_KEY_LENGTH, value_length);
+        for (key, value) in kvs {
+            modifies.push(Modify::Put(CF_DEFAULT, Key::from_raw(&key), value))
+        }
+    }
+    let ctx = Context::new();
+    let _ = engine.async_write(&ctx, modifies, Box::new(move |(_, _)| {}));
+}
+
+/// Measuring the performance of Engine::snapshot()
+fn engine_snapshot_bench<E: Engine>(
```

Comment:
How about just remove this function and put its content inside `bench_engine_snapshot`? It should be more easy to read.